### PR TITLE
WIP: Add _plot_legend for spatial colors bf plot

### PR DESCRIPTION
**Pull request by jona-sassenhagen**
screen shot 2015-11-30 at 22 09 30

### CODE REVIEW

**Comment by Eric89GXL** [comment]
Looks cool!

**Comment by jona-sassenhagen** [comment]
If Travis passes, review please? It's not much.
I copied a bunch of code from the topoplot code, and I'm not entirely sure I understand all of it.

**Comment by Eric89GXL** [comment]
> I copied a bunch of code from the topoplot code, and I'm not entirely sure I understand all of it.

If possible please refactor to reuse code instead of copying to keep things DRY.

**Comment by agramfort** [comment]
@jaeilepp please review.

File: mne/viz/evoked.py [hunk 1]
```diff
@@ -253,6 +283,8 @@ def _plot_evoked(evoked, picks, exclude, unit, show,
                         locs3d = np.array([ch['loc'][:3] for ch in chs])
                         x, y, z = locs3d.T
                         colors = _rgb(x, y, z)
+                        pos = find_layout(evoked.info).pos[:, :2][idx]
```
Comment (jaeilepp) [inline mne/viz/evoked.py]: I get an ``IndexError`` here.

Comment (jaeilepp) [inline mne/viz/evoked.py]: It's because ``t = 'eeg'`` and ``find_layout`` only returns meg channels.

File: mne/viz/evoked.py [hunk 2]
```diff
@@ -126,6 +127,35 @@ def _rgb(x, y, z):
     return np.asarray([x, y, z]).T
 
 
+def _plot_legend(pos=None, colors=None, axis=None, outlines='head'):
+    """Helper function to plot color/channel legends for butterfly plots
+    with spatial colors"""
+    from mpl_toolkits.axes_grid.inset_locator import inset_axes
+    ax = inset_axes(axis, width=0.5, height=0.5, loc=3)
```
Comment (jaeilepp) [inline mne/viz/evoked.py]: And here I get ``AttributeError: 'AnchoredSizeLocator' object has no attribute '_child'``

File: mne/viz/evoked.py [hunk 3]
```diff
@@ -126,6 +127,35 @@ def _rgb(x, y, z):
...
+
+    ax.set_xticks([])
+    ax.set_yticks([])
+    ax.set_frame_on(False)
+    if any([not pos_y.any(), not pos_x.any()]):
+        raise RuntimeError('No position information found, cannot compute '
+                           'geometries for topomap.')
+
+    for x, y, c in zip(pos_x, pos_y, colors):
+        ax.add_artist(patches.Circle(xy=(x, y), radius=0.03, color=c))
```
Comment (jaeilepp) [inline mne/viz/evoked.py]: Where is ``patches`` defined?

File: mne/viz/evoked.py (hunk 2 above)
Comment (jaeilepp) [inline mne/viz/evoked.py]: This might be a backend problem on my computer.

File: mne/viz/evoked.py (hunk 3 above)
Comment (jona-sassenhagen) [inline mne/viz/evoked.py]: My bad ... accidentally deleted the import. I don't understand the AttributeError. What context does it appear in?

**Comment by jaeilepp** [comment]
The legend seems to overlap with the loading text if you paint an area. I know it can be painful to position the text items in matplotlib, but I wonder if it would be easier to locate the legend to bottom right corner of the plot.

**Comment by jaeilepp** [comment]
Or maybe top left. I'm trying to think where it's overlapping data the least.

**Comment by jona-sassenhagen** [comment]
What loading text?

Repositioning shouldn't be that hard, as it's done by the `loc` parameter to inset_axes. Works like the one to `legend` IIRC.

**Comment by jaeilepp** [comment]
You can paint an area in the plot with left mouse button. Give it a try!

**Comment by jona-sassenhagen** [comment]
Ah ... I never use interactive plots :)

I think bottom right would be better than top.

**Comment by jona-sassenhagen** [comment]
The thing with lower left is that it's the position you least expect effects to appear in cause it's the baseline.

Top left I don't like because it visually interferes with the joint topoplot I like to use these for.

**Comment by jaeilepp** [comment]
Yeah. Maybe we should move the loading text instead. That could actually even appear in the center, cause it's there only for a short period of time.

**Comment by jona-sassenhagen** [comment]
If that would be possible ...

Could you maybe do that when you eventually take over this PR? I don't know the interactive plotting code at all.

**Comment by jaeilepp** [comment]
Yeah, I can do it.

File: mne/viz/evoked.py [hunk 4]
```diff
@@ -126,6 +127,36 @@ def _rgb(x, y, z):
     return np.asarray([x, y, z]).T
 
 
+def _plot_legend(pos=None, colors=None, axis=None, outlines='head'):
```
Comment (jaeilepp) [inline mne/viz/evoked.py]: I don't think you need to have default values here since you don't have checks for them.

Comment (jaeilepp) [inline mne/viz/evoked.py]: And I wonder if we even need the ``outlines`` param here. It doesn't seem to work very well in the legend.

Comment (jona-sassenhagen) [inline mne/viz/evoked.py]: IMO `outlines='skirt'` looks better than `'head'`, but elsewhere, we give the option of selecting either.

Not sure about the API, but I'd like to keep the option available at least here.

Comment (jona-sassenhagen) [inline mne/viz/evoked.py]: BTW channel type selection be fixed now.

**Comment by Eric89GXL** [comment]
You could also put the loading text on top with a higher z order and it
might look fine. The loading text is temporary so it might work decently.

**Comment by jona-sassenhagen** [comment]
When is it supposed to disappear by the way? Cause I don't see it disappearing right now. Neither is the selection. (Mac OSX)

**Comment by jaeilepp** [comment]
That's weird. There shouldn't be anything backend specific going on with the text item.

**Comment by jaeilepp** [comment]
Can I take over now?

**Comment by jaeilepp** [comment]
``zorder`` wont work because it's a separate axis.

**Comment by jona-sassenhagen** [comment]
Yes, take over.
Consider the 'skirt' option for outlines, IMO it's nicer.

**Comment by jona-sassenhagen** [comment]
Close for #2662

**Pull request closed without merging**